modules/p2/ding_list/theme/theme.inc

```diff
@@ -269,21 +267,33 @@ function template_preprocess_ding_list_add_button(&$variables) {
   }
 
   if ($variables['type'] == 'ting_object') {
-    $new_list = array(
-      'custom' => array(
-        'title' => '<div class="list-title">' . t('New list') . '</div><div class="add-to-list-button button">' . t('Add to new list') . '</div>',
-        'href' => 'list/create',
-        'attributes' => array(
-          'class' => array(),
+    if (user_access('use personalisation')) {
+      $new_list = array(
+        'custom' => array(
+          'title' => '<div class="list-title">' . t('New list') . '</div><div class="add-to-list-button button">' . t('Add to new list') . '</div>',
+          'href' => 'list/create',
+          'attributes' => array(
+            'class' => array(),
+          ),
+          'query' => array('at' => $variables['type'], 'av' => $variables['value']),
+          'html' => TRUE,
         ),
-        'query' => array('at' => $variables['type'], 'av' => $variables['value']),
-        'html' => TRUE,
-      ),
-    );
+      );
+    }
+    elseif (!user_is_logged_in()) {
+      $new_list = array(
+        'custom' => array(
+          'title' => '<div class="list-title">' . t('Login before you can use lists') . '</div><div class="add-to-list-button button">' . t('Login') . '</div>',
```

Comment:
Is there a reason why the existing approach with `ajax_command_ding_user_authenticate()` is not used here? It should provide a smoother user experience where a login automatically followed by the original intent of the user.

See [`ding_reservation_reserve_ajax()`](https://github.com/ding2/ding2/blob/50f3a749207ddd120772237f1fe3119afd0cee45/modules/ding_reservation/ding_reservation.module#L149-L173) for an example of how.

Reply:
At this point we don't know the user, so we don't know which list they want the object added to.
Not using the ajax call saves us a back and forth between the server, when we client side already knows what to do.

Reply:
I see the issue if this element is supported to show an entry for each list the item can be added to then we cannot predict at this point which of these the user intends to click next.

The preexisting solution would be to add a single "Add to list" button to elements and then (after login if needed) the user would choose the list in a modal or the like. I guess that ship has sailed.

If we are to do things differently I would prefer a more widely usable solution e.g. just adding a single link "Add to new list" with a specific class. If a link/button with such a class is clicked and the user is not logged in this will cause the login modal to open and if login succeeds this redirects to the original target.

Otherwise I fear that this pattern of handling the unauthenticated user will spread throughout the codebase.